Add token-based authentication

StefanPalashev opened the pull request:
# 🔐 Add Token-Based Authentication with Azure Entra ID Support for Redis Connections

## 📋 Summary

This Pull Request introduces comprehensive token-based authentication support, with first-class integration for Azure Entra ID (formerly Azure Active Directory) for Redis connections, enabling secure authentication for Azure-hosted Redis instances and Redis Enterprise with Azure integration.

## ✨ Features Added

### 🔑 **Multiple Authentication Flows**
- **DeveloperToolsCredential** - Authenticates through developer tools such as the Azure CLI.
It tries the following credential types, in this order, stopping when one provides a token:
    * `AzureCliCredential`
    * `AzureDeveloperCliCredential`
- **Service Principal** - `Client secret` and `certificate-based` authentication
- **Managed Identity** - Both `system-assigned` and `user-assigned`
- **Custom Credentials** - Support for any `TokenCredential` implementation

### **Token Management**
- **Automatic Token Refresh & Streaming of Credentials** - Seamlessly handle token expiration and stream updated credentials to prevent connection failures due to token expiration
- **Configurable Refresh Policies** - Customizable refresh thresholds and retry behavior

### 🛡️ **Input Validation**

- **Empty Scope Prevention** - Rejects empty or invalid scope configurations
- **Scope Validation** - Ensures that scopes are not empty. 
The validation can be extended to follow Azure's `https://resource/.default` format.

## 🏗️ **Architecture**

### **Async-First Design**
- Full async/await support for non-blocking operations
- Seamless integration with multiplexed connections

### **Streaming Credentials Pattern**
The implementation uses a streaming approach where:
1. The credentials provider (`EntraIdCredentialsProvider`) runs a background task for token refresh
2. Connections subscribe to credential updates via async streams
3. Automatic re-authentication occurs when new tokens are available
4. Token refresh occurs without interrupting the connection

### **Error Handling**
The library provides detailed error messages for troubleshooting and properly handles authentication and token refresh failures. These messages are emitted via the `log` crate, allowing observation of the library’s runtime behavior through configured loggers.
The background token refresh service will automatically retry failed token refreshes according to the retry configuration, logging information about each attempt at appropriate levels.
Once the maximum number of attempts is exceeded, the service will stop retrying and the underlying error will be propagated to the subscribers, which are `multiplexed connections`, rendering them unusable.

### **Feature Flags**
- `token-based-authentication`: General traits and types for token-based authentication support
- `entra-id`: Azure Entra ID specific implementation includes the main dependencies - `azure_identity` and `azure_core`

### **New Modules**
```
redis/
├── src/
│   ├── auth.rs    # Core authentication traits and types
│   ├── auth_management.rs    # Token refresh configuration and utilities
│   └── entra_id.rs    # Azure Entra ID credentials provider implementation
└── tests/
    ├── test_acl.rs    # Integration tests using ACL and a mocked streaming credentials provider
    └── test_auth.rs    # Integration tests
    └── test_credentials_provider_failures.rs    # Tests for connection-layer behavior when streaming credentials providers fail.
```

### **Enhanced Modules**
```
redis/
└── src/
    ├──client.rs    # Attach a credentials provider to a client via the `with_credentials_provider` function
    ├──connection.rs    # Credentials provider in the `RedisConnectionInfo` struct
    ├──lib.rs    # New public APIs
    └── aio/
       └── multiplexed_connection.rs    # Streaming credentials subscription and re-authentication
```

## 🚀 **Usage Examples**

### **Basic Usage**
```rust
use redis::{Client, EntraIdCredentialsProvider, RetryConfig};

// Using DeveloperToolsCredential
let mut provider = EntraIdCredentialsProvider::new_developer_tools()?;
provider.start(RetryConfig::default());

let client = Client::open_with_credentials_provider("redis://your-redis.azure.com:6380", provider).unwrap();

let mut connection = client.get_multiplexed_async_connection().await?;
```

### **Service Principal Authentication**
```rust
use redis::{Client, EntraIdCredentialsProvider, RetryConfig};

// Using ClientSecretCredential
let mut provider = EntraIdCredentialsProvider::new_client_secret(
    "tenant-id".to_string(),
    "client-id".to_string(), 
    "client-secret".to_string(),
)?;
provider.start(RetryConfig::default());

let client = Client::open_with_credentials_provider("redis://your-redis.azure.com:6380", provider).unwrap();

let mut connection = client.get_multiplexed_async_connection().await?;
```

### **Managed Identities**
```rust
use redis::{Client, EntraIdCredentialsProvider, RetryConfig};

let mut provider = EntraIdCredentialsProvider::new_system_assigned_managed_identity()?;
// or
/* 
User assigned managed identities can be identified by one of the values in the following enumeration:
pub enum UserAssignedId {
    /// The client ID of a user-assigned identity
    ClientId(String),
    /// The object or principal ID of a user-assigned identity
    ObjectId(String),
    /// The Azure resource ID of a user-assigned identity
    ResourceId(String),
}
*/

let provider = EntraIdCredentialsProvider::new_user_assigned_managed_identity_with_scopes(
    vec!["your-scope".to_string()],
    Some(ManagedIdentityCredentialOptions {
        // Specify the user-assigned identity using one of:
        user_assigned_id: Some(UserAssignedId::ClientId("your-client-id".to_string())),
        // or: user_assigned_id: Some(UserAssignedId::ObjectId("your-object-id".to_string())),
        // or: user_assigned_id: Some(UserAssignedId::ResourceId("your-resource-id".to_string())),
        ..Default::default()
    }),
)?;
provider.start(RetryConfig::default());

let client = Client::open_with_credentials_provider("redis://your-redis.azure.com:6380", provider).unwrap();

let mut connection = client.get_multiplexed_async_connection().await?;
```

## 🧪 **Testing**

### **Comprehensive Test Suite**
- Unit tests for token refresh configuration
- Unit tests for scopes validation
- JWT parsing and validation tests
- Mock credential provider tests simulating various scenarios:
  - Successful authentication
  - Authentication failures
  - Retry mechanisms
  - Multiple token refreshes over time
  - Multiple subscribers
  - Provider cleanup
- Mocked streaming credentials provider tests with ACL, validating automatic re-authentications and connection stability
- Tests for connection-layer behavior when streaming credentials providers fail

### **Mock Infrastructure**
- `MockStreamingCredentialsProvider` and `MockTokenCredential` for testing without real Azure dependencies
- Configurable response patterns (success, failure, alternating)
- Call counting and verification

### **Integration Tests**
The integration tests are designed to ensure correct interaction with real Azure services and identity endpoints.

The tests rely on standard Azure environment variables to authenticate against Azure Entra ID:
* `AZURE_TENANT_ID`
* `AZURE_CLIENT_ID`
* `AZURE_CLIENT_SECRET`
* `AZURE_CLIENT_CERTIFICATE_PATH`

## 📦 **Dependencies**

### **New Dependencies**
```toml
[dependencies]
azure_identity = { version = "0.31", optional = true, features = ["client_certificate"] }
azure_core = { version = "0.31", optional = true }
base64 = { version = "0.22.1", optional = true }

[features]
token-based-authentication = ["dep:futures-util", "dep:log", "dep:backon"]
entra-id = ["dep:azure_identity", "dep:azure_core", "dep:base64", "token-based-authentication", "tokio-comp"]
```

### **Feature Gating**
All Entra ID functionality is properly feature-gated with `#[cfg(feature = "entra-id")]` to avoid dependency bloat for users who don't need Azure authentication.

## 🔧 **Configuration**

### **Cargo.toml**
```toml
[dependencies]
redis = { version = "1.0", features = ["entra-id", "tokio-comp"] }
```

## 🛡️ **Security Considerations**

- **Token Security** - Tokens are handled securely
- **Scope Validation** - Prevents misconfiguration that could lead to authentication failures
- **Error Handling** - Azure-specific errors are properly mapped to Redis errors.

## 📚 **Documentation**

### **Added Documentation**
- `redis/docs/token_authentication.md` - Detailed usage guide with examples and integration patterns for different deployment scenarios
- `README.md` - Updated with token-based authentication examples

## **Backward Compatibility**

- **Fully backward compatible** - No changes to existing APIs
- **Optional feature** - Only affects builds that explicitly enable `entra-id`
- **No breaking changes** - All existing authentication methods continue to work

## 🔍 **Code Quality**

- **Zero Code Duplication** - Common patterns extracted into reusable functions
- **Error Handling** - All error paths properly handled with descriptive messages
- **Memory Efficiency** - Minimal allocations and efficient token caching

## 🚦 **Testing Instructions**
The integration tests are ignored by default, as they are designed to be executed only against live environments.
They can be run manually using the following command:

```bash
cargo test --test test_auth --features entra-id -- --ignored
```
To run a specific test, the following command can be used:

```bash
cargo test --test test_auth --features entra-id <test_name> -- --ignored
```

## 🎉 **Benefits**

- **🔐 Enterprise Security** - Leverage Azure's enterprise-grade authentication
- **🔄 Zero-Downtime Authentication** - Automatic token refresh prevents connection drops
- **📈 Scalability** - Supports high-throughput applications with efficient token management
- **🛠️ Developer Experience** - Simple, intuitive API with examples
- **🏢 Enterprise Ready** - Supports all Azure authentication patterns used in production

## **Contact / Support**
For any future questions, feedback, or discussion related to this feature, members of the open source community can reach out at [ecosystems@redis.com](mailto:ecosystems@redis.com)

nihohit commented on redis/docs/token_authentication.md:
```diff
@@ -0,0 +1,217 @@
+- `initial_delay`: Initial delay before first retry
+- `max_delay`: Maximum delay between retries
+- `backoff_multiplier`: Exponential backoff multiplier
+- `jitter_percentage`: Random jitter percentage (0.0-1.0)
```
clarify what these retries mean

nihohit commented on redis/docs/token_authentication.md:
````diff
@@ -0,0 +1,217 @@
+
+```rust
+let provider = EntraIdCredentialsProvider::new_default()?;
+```
````
all of the code documentation should also (or only) appear in code, where it will be tested by rust doc tests

nihohit commented on redis/docs/token_authentication.md:
```diff
@@ -0,0 +1,217 @@
+
+The library provides comprehensive error handling for authentication and token refresh failures.
+The background token refresh service will automatically retry failed token refreshes according to the retry configuration.
+Once the maximum number of attempts is reached, the service will stop retrying and the underlying error will be propagated to the subscribers.
```
how & when will it propagate?

nihohit commented on redis/docs/token_authentication.md:
```diff
@@ -0,0 +1,217 @@
+### 2. Configure Appropriate Refresh Ratios
+
+- **High-frequency apps**: 0.5-0.7 (refresh early)
+- **Low-frequency apps**: 0.8-0.9 (refresh later)
```
what's a high/low-frequency app?

nihohit commented on redis/tests/test_auth.rs:
```diff
@@ -0,0 +1,397 @@
+#![cfg(feature = "token-based-authentication")]
```
these tests are all unit tests, and should be in the relevant files, not with external integration tests.

what you're missing is actual tests that check that connections get new credentials, and send these new credentials to the server.
You should create a test credentials provider, that doesn't rely on entra, and then either check that it refreshes the credentials with a mock server, or by spawning a task that periodically* changes the ACL password, and seeing that the connection is still usable.

*periodically here means every couple of milliseconds, because we don't want a test that runs for minutes or more.

nihohit commented on config/redis.toml:
```diff
@@ -99,6 +99,28 @@ require = ["smol-rustls-comp"]
+[[rule]]
+when = ["entra-id", "AND", ["tls-rustls", "OR", "tls-rustls-webpki-roots"]]
+require = ["tokio-rustls-comp"]
+forbid = ["NOT", "tokio-rustls-comp"]
```
this can be removed

nihohit commented on config/redis.toml:
```diff
@@ -99,6 +99,28 @@ require = ["smol-rustls-comp"]
+[[rule]]
+when = ["entra-id", "AND", "tls-native-tls"]
+require = ["tokio-native-tls-comp"]
+forbid = ["NOT", "tokio-native-tls-comp"]
```
same

nihohit commented on redis/src/auth_management.rs:
```diff
@@ -125,3 +125,74 @@ pub mod credentials_management_utils {
 }
+
+#[cfg(all(feature = "token-based-authentication", test))]
+mod auth_management_tests {
```
since the tests are here, then things that shouldn't be `pub` can be just `pub(crate)` or even private in the file above. please ensure everything is minimally public.

nihohit commented on redis/docs/token_authentication.md:
```diff
@@ -0,0 +1,225 @@
+# Token-Based Authentication with Azure Entra ID
```
instead of adding this as a separate document in a new folder, let's try to integrate this into the library's rustdoc, in a way that will use our ability to check that the code actually compiles. Please move this to the `entra_id` module, and check the generated rust docs to see that it looks like you expect the doc to look, and that each named type is linked to the type's documentation.

nihohit commented on redis/docs/token_authentication.md:
```diff
@@ -0,0 +1,225 @@
+- `initial_delay`: Initial delay before attempting to refresh the token after a failure. Subsequent retries use exponential backoff based on this value.
+- `max_delay`: Upper bound for retry delays to prevent excessively long waits. Delays will never exceed this value, even with exponential backoff.
+- `backoff_multiplier`: Growth factor for exponential backoff (typically 2.0 for doubling). Each retry delay is multiplied by this value: `delay = delay * backoff_multiplier`.
+- `jitter_percentage`: Random variation added to delays as a fraction of the calculated delay (0.0 to 1.0). For example, 0.5 means up to ±50% variation to prevent synchronized retries.
```
once you integrate this into the rustdoc, there won't be a need to re-document each type, just link to the type's documentation.

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -587,6 +633,36 @@ impl MultiplexedConnection {
+                        }
+                    }
+                }
+                println!("Re-authentication stream ended");
```
please remove these prints. We should consider how to pass failures back to the user, but the console isn't the right place for it.

nihohit replied:
terminate the connection.

nihohit commented on redis/src/auth.rs:
```diff
@@ -0,0 +1,65 @@
+/// This trait allows different authentication mechanisms to be plugged into
+/// the Redis client, such as static passwords, token-based authentication,
+/// or dynamic credential providers like Azure Entra ID.
+pub trait CredentialsProvider: Send + Sync {
```
I don't see where this, or StaticCredentialsProvider, are used.

nihohit commented on redis/src/entra_id.rs:
```diff
@@ -0,0 +1,951 @@
+    pub password: Option<String>,
+}
+
+type Subscriptions = Vec<Arc<Sender<RedisResult<BasicAuth>>>>;
```
I don't think that the `Arc` here is necessary, the sender is already cheaply clonable.

nihohit commented on redis/src/entra_id.rs:
```diff
@@ -0,0 +1,951 @@
+    pub fn start(&mut self, retry_config: RetryConfig) {
+        self.start_refresh_service(retry_config, |access_token| {
+            let remaining = access_token.expires_on - OffsetDateTime::now_utc();
+            let remaining_duration = match remaining.try_into() {
```
what's the type conversion here? are you converting duration to duration?

nihohit commented on redis/src/entra_id.rs:
```diff
@@ -0,0 +1,951 @@
+                Err(_) => std::time::Duration::from_secs(0),
+            };
+            remaining_duration
+                .checked_sub(std::time::Duration::from_secs(240))
```
why is this a constant value?

nihohit replied:
then please put it under a named, commented constant.

StefanPalashev replied:
The value has been extracted to a constant.

StefanPalashev commented:
Thanks for the review! I’m working through the comments now and will follow up once everything is addressed.

StefanPalashev commented on redis/src/entra_id.rs:
```diff
@@ -0,0 +1,951 @@
+        if self
+            .background_handle
+            .lock()
+            .expect("mutex poisoned")
```
The code has been replaced with an `unwrap()`.

StefanPalashev commented on redis/src/entra_id.rs:
```diff
@@ -0,0 +1,951 @@
+        }))
+        .await;
+
+        subscribers_arc
```
Thanks for the suggestion!
The code has been modified to reflect it. As a result, the necessary cleanup is performed upfront, ensuring that only active subscribers are being notified about the messages.

StefanPalashev commented on redis/docs/token_authentication.md:
```diff
@@ -0,0 +1,225 @@
+### 2. Handle Token Expiration
+
+- Use background refresh services for long-running applications
+- Implement proper error handling for authentication failures
```
These recommendations are no longer applicable, as the `StreamingCredentialsProvider` handles all of these cases internally. They were only relevant when alternative credentials providers could be used, which is no longer possible since the other traits have been removed.

As a result, users do not need to take any action—starting the credentials provider automatically covers these scenarios.

StefanPalashev commented:
@nihohit 
I’ve addressed the feedback from the previous review. When you have a moment, I’d appreciate another look and any guidance you may have on the remaining items (especially the ones targeting `redis/src/aio/multiplexed_connection.rs`).

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -585,7 +585,7 @@ impl MultiplexedConnection {
                 }
                 None => {
-                    println!("Credential stream closed");
+                    unreachable!("Credential stream closed");
```
if there's no possible way to output that this happens without crashing the user program, it's better to keep this codepath empty.
Just do
```
None => {
  // This path is taken when the credential stream is closed. ATM we do nothing - 
  //  the user will notice this issue if / when the connection will stop authenticating.
}

nihohit commented on redis/src/entra_id.rs:
```diff
@@ -245,7 +413,10 @@ impl EntraIdCredentialsProvider {
                             retry_config.max_delay,
                         );
-                        println!("An error occurred while refreshing the token. Attempt {attempt}. Sleeping for {:?}", error_delay);
+                        println!(
```
again, remove printlns from the code. If you want, instead use the `log` dependency in the feature.

StefanPalashev replied:
The `println!` calls were replaced with logging via the `log` crate.

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -558,9 +567,32 @@ impl MultiplexedConnection {
+                    connection_info.password = Some(ArcStr::from(credentials.password));
+                }
+                Some(Err(err)) => {
+                    eprintln!("Error while receiving credentials from stream: {err}");
```
use the `log` crate

StefanPalashev replied:
The `log` crate has been introduced as a dependency for the feature.

nihohit commented on redis/src/entra_id.rs:
```diff
@@ -0,0 +1,1164 @@
+                        tokio::time::sleep(error_delay).await;
+                        continue;
+                    }
+                    println!("Max attempts reached. Stopping token refresh.");
```
use the `log` crate

StefanPalashev replied:
The `log` crate has been introduced as a dependency for the feature.

nihohit commented:
@StefanPalashev I marked as resolved the Copilot comments that I found irrelevant.

StefanPalashev commented on redis/src/entra_id.rs:
```diff
@@ -0,0 +1,1164 @@
+                        ErrorKind::AuthenticationFailed,
+                        "Failed to unwrap credential",
+                    ))
+                })?,
```
The code was extracted into a helper function.

StefanPalashev commented on redis/src/entra_id.rs:
```diff
@@ -0,0 +1,951 @@
+                std::sync::Arc::try_unwrap(credential_provider).map_err(|_| {
+                    RedisError::from((
+                        ErrorKind::AuthenticationFailed,
+                        "Failed to unwrap credential",
```
This code was duplicated across multiple constructors. It is now a separate function with a more descriptive message.

StefanPalashev commented on README.md:
```diff
@@ -163,6 +163,36 @@ To enable insecure mode, append `#insecure` at the end of the URL:
+- **DeveloperToolsCredential**: Authenticates through developer tools such as the Azure CLI.
+It tries the following credential types, in this order, stopping when one provides a token:
+* [`AzureCliCredential`]
+* [`AzureDeveloperCliCredential`]
```
This suggestion was ignored.

StefanPalashev commented:
As discussed above (in https://github.com/redis-rs/redis-rs/pull/1831#discussion_r2684022129), the connection has to be terminated in the event of re-authentication errors or errors caused by the credentials provider. However, the latest changes introduce a mechanism that does not terminate the connection, but instead leaves it in a practically unusable state by causing all subsequent commands to fail with a re-authentication error.

Is the correct way to terminate the connection to drop the Pipeline’s `mpsc::Sender` so that the channel closes and the driver task is aborted?

@nihohit, could you please provide guidance on how to approach this, as the current solution is probably not sufficient enough? Thanks in advance!

nihohit commented:
> Is the correct way to terminate the connection to drop the Pipeline’s mpsc::Sender so that the channel closes and the driver task is aborted?

That will only work if all the clones of the connection did this, and I don't think that this is enforceable. I think that if the authentication stream terminates and an error is logged, then the rest of the failed requests will be clear.

nihohit commented on .github/workflows/rust.yml:
```diff
@@ -289,7 +289,15 @@ jobs:
         uses: ./.github/actions/lint-and-check
         with:
-          rust-toolchain: stable    
+          rust-toolchain: stable
```
revert

nihohit commented:
@StefanPalashev, sorry, pushed another commit with the fix to flag-frenzy. once you pull that and answer https://github.com/redis-rs/redis-rs/pull/1831#discussion_r2777883459, I think that we'll be done :)

StefanPalashev commented:
> > Is the correct way to terminate the connection to drop the Pipeline’s mpsc::Sender so that the channel closes and the driver task is aborted?
> 
> That will only work if all the clones of the connection did this, and I don't think that this is enforceable. I think that if the authentication stream terminates and an error is logged, then the rest of the failed requests will be clear. I guess we'll just have to rely on the error logs.

This should mean that we can invoke `drop` on the sender when `re_authentication_failed` has been set, eventually all of the clones will get their senders closed if they attempt to send commands.

nihohit commented:
> This should mean that we can invoke drop on the sender when re_authentication_failed has been set, eventually all of the clones will get their senders closed if they attempt to send commands.

Is this an answer to my last comment?
I was asking why the change is only for tokio - I couldn't find any reference that says entra is only supported by tokio.

nihohit commented:
Thanks for the contribution! I'll create a release soon.
